// colyseus.js/lib/room.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

exports.createRoom = createRoom;

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var EventEmitter = require('events').EventEmitter,
    jsonpatch = require('fast-json-patch'),
    msgpack = require('msgpack-lite'),
    fossilDelta = require('fossil-delta'),
    Clock = require('clock.js'),
    protocol = require('./protocol');

var Room = function (_EventEmitter) {
  _inherits(Room, _EventEmitter);

  function Room(client, name) {
    _classCallCheck(this, Room);

    var _this = _possibleConstructorReturn(this, (Room.__proto__ || Object.getPrototypeOf(Room)).call(this));

    _this.id = null;
    _this.client = client;

    _this.name = name;
    _this.state = {};

    _this.clock = new Clock();
    _this.remoteClock = new Clock();

    _this.lastPatchTime = null;
    _this.ping = null;

    _this.on('leave', function () {
      return _this.removeAllListeners();
    });

    return _this;
  }

  _createClass(Room, [{
    key: 'setState',
    value: function setState(state, remoteCurrentTime, remoteElapsedTime) {

      this.state = state;
      this._previousState = msgpack.encode(this.state);

      // set remote clock properties
      this.remoteClock.currentTime = remoteCurrentTime;
      this.remoteClock.elapsedTime = remoteElapsedTime;

      this.clock.start();

      this.emit('update', state);
    }
  }, {
    key: 'patch',
    value: function patch(binaryPatch) {

      //
      // calculate client-side ping
      //
      var patchTime = Date.now();

      if (this.lastPatchTime) {
        this.ping = patchTime - this.lastPatchTime;
      }

      this.lastPatchTime = patchTime;

      this.clock.tick();

      //
      // apply patch
      //
      this._previousState = fossilDelta.apply(this._previousState, binaryPatch);
      var newState = msgpack.decode(this._previousState);

      var patches = jsonpatch.compare(this.state, newState);
      this.emit('patch', patches);

      this.state = newState;
      this.emit('update', this.state, patches);
    }
  }, {
    key: 'leave',
    value: function leave() {

      if (this.id >= 0) {

        this.client.send([protocol.LEAVE_ROOM, this.id]);
      }
    }
  }, {
    key: 'send',
    value: function send(data) {

      this.client.send([protocol.ROOM_DATA, this.id, data]);
    }
  }]);

  return Room;
}(EventEmitter);

function createRoom(client, name) {

  return new Room(client, name);
}
